File: src/python/radix_sort.py
def radix_sort(arr):
    # Find the maximum number to know the number of digits
    max_num = max(arr)

    # Do counting sort for every digit, starting from the least significant digit
    exp = 1
    while max_num // exp > 0:
        counting_sort(arr, exp)
        exp *= 10


def counting_sort(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for i in range(n):
        index = arr[i] // exp
        count[index % 10] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1
        i -= 1

    for i in range(n):
        arr[i] = output[i]


def main():
    arr = [170, 2, 45, 75, 75, 90, 802, 24, 2, 66]
    print("Unsorted array:", arr)
    radix_sort(arr)
    print("Sorted array:", arr)


if __name__ == "__main__":
    main()

File: src/c/RadixSort.c
#include <stdio.h>

void print_array(int array[], int size) {
  printf("[ ");
  for (int i = 0; i < size; i++) {
    printf("%d ", array[i]);
  }
  printf("]\n");
}

int get_max(int array[], int size) {
  int max = array[0];
  for (int i = 1; i < size; i++) {
    if (array[i] > max)
      max = array[i];
  }
  return max;
}

void radix_sort(int array[], int size) {
  int i;
  int semi_sorted[size];
  int significant_digit = 1;
  int largest_number = get_max(array, size);

  while (largest_number / significant_digit > 0) {
    int bucket[10] = {0};
    for (i = 0; i < size; i++) {
      bucket[(array[i] / significant_digit) % 10]++;
    }
    for (i = 1; i < 10; i++) {
      bucket[i] += bucket[i - 1];
    }
    for (i = size - 1; i >= 0; i--) {
      semi_sorted[--bucket[(array[i] / significant_digit) % 10]] = array[i];
    }
    for (i = 0; i < size; i++) {
      array[i] = semi_sorted[i];
    }
    significant_digit *= 10;
  }
}

int main() {
  int array[10] = {45, 75, 89, 12, 34, 9, 67, 23, 90, 11};
  printf("Unsorted Array\n");
  print_array(array, 10);

  radix_sort(array, 10);

  printf("Sorted Array\n");
  print_array(array, 10);
  return 0;
}
